recirq/quantum_chess/circuit_transformer.py

```diff
@@ -305,6 +317,384 @@ def transform(self, circuit: cirq.Circuit) -> cirq.Circuit:
+                shortest[i][j] = 1
+        for k in range(v):
+            for i in range(v):
+                for j in range(v):
+                    shortest[i][j] = min(shortest[i][j], shortest[i][k] + shortest[k][j])
+
+        # For each node, find the length of the shortest path to the farthest
+        # node
+        farthest = [0 for i in range(v)]
+        for i in range(v):
+            for j in range(v):
+                if i != j and shortest[i][j] > farthest[i]:
+                    farthest[i] = shortest[i][j]
+
+        # Find the graph center such that the length of the shortest path to the
+        # farthest node is the smallest. Use the first graph center if there are
+        # multiple graph centers.
+        center = 0
+        for i in range(v):
+            if farthest[i] < farthest[center]:
+                center = i
+
+        return index_to_qubit_mapping[center]
+
+    def traverse(
+        self,
+        g: Dict[cirq.Qid, List[Tuple[cirq.Qid, int]]],
+        s: cirq.Qid,
+        visited: Optional[set] = None,
+    ) -> Deque[cirq.Qid]:
+        """Returns a deque of qubits ordered by breadth-first search traversal.
+
+        During each iteration of breadth-first search, the adjacent nodes are
+        sorted by their corresponding moments before being traversed.
+
+        Args:
+          g: A logical qubits graph.
+          s: The source qubit from which to start breadth-first search.
+        """
+        order = deque()
+        if visited is None:
+            visited = set()
+        visited.add(s)
+        queue = deque()
+        queue.append(s)
+        while queue:
+            q = queue.popleft()
+            order.append(q)
+            neighbors_sorted_by_moment = sorted(g[q], key=lambda x: x[1])
+            for neighbor in neighbors_sorted_by_moment:
+                neighbor = neighbor[0]
+                if neighbor not in visited:
+                    visited.add(neighbor)
+                    queue.append(neighbor)
+        return order
+
+    def find_reference_qubits(
+        self,
+        mapping: Dict[cirq.Qid, cirq.GridQubit],
+        lg: Dict[cirq.Qid, List[Tuple[cirq.Qid, int]]],
+        lq: cirq.Qid,
+    ) -> List[cirq.GridQubit]:
+        """Returns a list of physical qubits from which to find the next mapping.
+
+        The nodes adjacent to the logical qubit parameter are sorted by their
+        corresponding moments before being traversed. For each adjacent node
+        that has been mapped to a physical qubit, the mapped physical qubit is
+        added to the result.
+
+        Args:
+          mapping: The current mapping of logical qubits to physical qubits.
+          lg: A logical qubits graph.
+          lq: The logical qubit from which to find reference qubits.
+        """
+        qubits = []
+        neighbors_sorted_by_moment = sorted(lg[lq], key=lambda x: x[1])
+        for neighbor in neighbors_sorted_by_moment:
```

Comment:
nit, this can be simplified:
```
for neighbor in neighbors_sorted_by_moment.keys():
```

Reply:
`neighbors_sorted_by_moment` is a list but updated to make this a bit cleaner.